Adding proper GUI / code for custom model support (gui reimplementation from mcyoung)

Teotia444 opened the pull request:
This PR is a re-implementation of a re-implementation from McYoung that contains code for a custom model subpage. 

(from mcy's pr) The PR contains : 
  1. Code for doing the patching. Right now it patches Link.szs and some files for voice clips. It should probably also patch the files for the map icon, too, but I don't know what the right destination for those is.
  2. Adds a way to select a custom model in the Aroma UI. The current mechanism is not ideal or efficient, but unless people have hundreds of models I doubt the O(n log n) overhead will be noticeable. I have made it so that you can have multiple models to choose from.
This PR still doesn't contain the desktop code, but i'll probably open another one containing just that.

Very quick video showing how the added tabs look : 
https://github.com/user-attachments/assets/473a9408-056b-4dca-a6b9-efd0b420c6e3
(can provide more if needed)

Teotia444 committed:
custom models experimental

Teotia444 committed:
Update README.md

Teotia444 committed:
Update README.md

Teotia444 committed:
Update README.md

Teotia444 committed:
Update README.md

Teotia444 committed:
Update README.md

Teotia444 committed:
branching from mcy

Teotia444 committed:
Merge branch 'SuperDude88:main' into main

Teotia444 committed:
custom models tab

Teotia444 committed:
Hide color related features when selecting custom models

Teotia444 committed:
Custom models cleanup

Teotia444 committed:
Quick checks for custom model implementation

mcy commented on customizer/model.cpp:
```diff
@@ -248,8 +249,122 @@ static const std::unordered_map<std::string, std::list<std::string>> casualTextu
 // IMPROVEMENT: Better generalize this in the future
 ModelError CustomModel::applyModel() const {
+    //start by counting the folders in case the user decided to select a random model even though theres no model
+    int fileCount = std::count_if(
+        begin(std::filesystem::directory_iterator(Utility::get_models_dir())),
+        end(std::filesystem::directory_iterator(Utility::get_models_dir())),
+        [](auto& entry) { return entry.is_directory(); }
+    );
```
```cc
std::vector<fspath> models;
for (auto& dirent : std::filesystem::directory_iterator(Utility::get_models_dir())) {
  if (dirent.is_directory()) models.push(dirent.path());
}
std::sort(models.begin(), models.end()); // readdir order isn't deterministic
```

Teotia444 replied:
thanks, much cleaner!

mcy commented on customizer/model.cpp:
```diff
@@ -248,8 +249,122 @@ static const std::unordered_map<std::string, std::list<std::string>> casualTextu
+        Utility::platformLog("Applying custom model " + modelName + "...");
+
+        auto model = Utility::get_models_dir() / modelName;
+        if (modelName == "random") {
```
Perhaps this should not overload the value of modelName, since people will get confused if they name a model random...

Teotia444 replied:
I've replaced "random" with "?" as it is a forbidden character in Windows, which is already less likely to cause issues. I agree tho, modelName should probably not be overloaded and i should just add a new attribute, i'm lazy tho :)

Teotia444 committed:
Cleaned up custom models v2

cleaner code for model fetcher and fixed potential issue with overloading modelName

Teotia444 commented:
gonna close this for now to clean up my branches, if this is still useful i'm keeping a branch of this but yeah
